### PR TITLE
Feature/add delete endpoint

### PR DESCRIPTION
**Pull request by southeo**
Added functionality to delete mappings and source systems from the database. Records are not removed from the DB; instead, their "deleted" field is updated. 

In paginated calls, we ignore deleted mappings/source systems, but we can still search for them by specific id

[Jira for this PR](https://naturalis.atlassian.net/browse/DD-323?atlOrigin=eyJpIjoiZDYxZGMwZjQ5MGE1NGM3ZGE4OTg1YWE2ZWU0Y2VhMTAiLCJwIjoiaiJ9)

Next: Tombstone PIDs of deleted mappings and source systems. [Jira ticket](https://naturalis.atlassian.net/browse/DD-323?atlOrigin=eyJpIjoiYmJiODE2YWExZTBhNGQwNDgyMzEzZGQ4YjQyODBhYWIiLCJwIjoiaiJ9)

### CODE REVIEW

File: src/main/java/eu/dissco/orchestration/backend/controller/MappingController.java [hunk 1]
```diff
@@ -48,13 +49,24 @@ public ResponseEntity<MappingRecord> createMapping(Authentication authentication
...
     log.info("Received update request for mapping: {}", id);
-    var result = service.updateMapping(id, mapping, getNameFromToken(authentication));
+    MappingRecord result = null;
+    result = service.updateMapping(id, mapping, getNameFromToken(authentication));
     return ResponseEntity.ok(result);
   }
 
+  @PreAuthorize("isAuthenticated()")
+  @ResponseStatus(HttpStatus.NO_CONTENT)
+  @PatchMapping(value = "/{prefix}/{suffix}", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
```
Comment (samleeflang) [inline src/main/java/eu/dissco/orchestration/backend/controller/MappingController.java:62]: Can we change this to Delete.
Seems like a relatively small change

File: src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java [hunk 2]
```diff
@@ -47,28 +48,53 @@ public MappingRecord getMapping(String id) {
         .fetchOne(this::mapToMappingRecord);
   }
 
+  public MappingRecord getMappingOmitDeleted(String id) {
```
Comment (samleeflang) [inline src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java:51]: Yes let's leave this as is for now. Maybe later we need to remove versions for this from the database as well.

File: src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java [hunk 3]
```diff
@@ -47,28 +48,53 @@ public MappingRecord getMapping(String id) {
...
 
-  private MappingRecord mapToMappingRecord(Record record) {
+  public int mappingExists(String id){
+    return context.select(NEW_MAPPING.ID)
+        .from(NEW_MAPPING)
+        .where(NEW_MAPPING.ID.eq(id))
+        .and(NEW_MAPPING.DELETED.isNull())
+        .fetch().size();
+  }
+  public void deleteMapping(String id, Instant deleted){
```
Comment (samleeflang) [inline src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java:79]: Sorry minor thing, a newline between the two methods

File: src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java (hunk 2 above)
Comment (samleeflang) [inline src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java:51]: I would just call this `getMapping` or if you want `getActiveMapping` this seems a bit long. Or add get `latest` in it.

File: src/main/java/eu/dissco/orchestration/backend/service/MappingService.java [hunk 4]
```diff
@@ -30,8 +31,12 @@ public MappingRecord createMapping(Mapping mapping, String userId) throws Transf
     repository.createMapping(mappingRecord);
     return mappingRecord;
   }
-  public MappingRecord updateMapping(String id, Mapping mapping, String userId) {
-    var currentVersion = repository.getMapping(id);
+  public MappingRecord updateMapping(String id, Mapping mapping, String userId)
+      throws NotFoundException {
+    var currentVersion = repository.getMappingOmitDeleted(id);
+    if (currentVersion == null){
+      throw new NotFoundException("Requested mapping does not exist");
```
Comment (samleeflang) [inline src/main/java/eu/dissco/orchestration/backend/service/MappingService.java:38]: Maybe at the id of the requested mapping in the exception. Could help debugging, if it goes wrong and we should have returned a mapping we can check the id with what is in the database.

File: src/main/java/eu/dissco/orchestration/backend/service/MappingService.java [hunk 5]
```diff
@@ -42,6 +47,15 @@ public MappingRecord updateMapping(String id, Mapping mapping, String userId) {
     }
   }
 
+  public void deleteMapping(String id) throws NotFoundException {
+    int result = repository.mappingExists(id);
+    if (result > 0){
+      Instant deleted = Instant.now();
+      repository.deleteMapping(id, deleted);
```
Comment (samleeflang) [inline src/main/java/eu/dissco/orchestration/backend/service/MappingService.java:54]: Feels a bit strange to have a different method here than in update for checking if the mapping exists. Might make these the same, and only check latest version (same as update). If we delete we always delete all.

File: src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java [hunk 6]
```diff
@@ -47,28 +48,53 @@ public MappingRecord getMapping(String id) {
...
   }
 
+  public MappingRecord getMappingOmitDeleted(String id) {
+    return context.select(NEW_MAPPING.asterisk())
+        .distinctOn(NEW_MAPPING.ID)
+        .from(NEW_MAPPING)
+        .where(NEW_MAPPING.ID.eq(id))
+        .and(NEW_MAPPING.DELETED.isNull())
+        .orderBy(NEW_MAPPING.ID, NEW_MAPPING.VERSION.desc())
+        .fetchOne(this::mapToMappingRecord);
```
Comment (samleeflang) [inline src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java:58]: Could use `fetchOptional()` here, might be nicer

**Review by samleeflang — COMMENTED** [review]
Couple of small comments.
Discussed this afternoon to make them DELETE after all, sorry for the confusion.

File: src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java (hunk 2 above)
Comment (southeo) [reply src/main/java/eu/dissco/orchestration/backend/repository/MappingRepository.java:51]: `getActiveMapping` is most accurate. 
It needs to be distinct from `getMapping`, which may or may not return a deleted mapping (as we discussed, we might want to do this if the user specifically requests it.)

File: src/main/java/eu/dissco/orchestration/backend/service/SourceSystemService.java [hunk 7]
```diff
@@ -48,6 +53,15 @@ public JsonApiWrapper getSourceSystemRecords(int pageNum, int pageSize, String p
     return wrapResponse(sourceSystemRecords, pageNum, pageSize, path);
   }
 
+  public void deleteSourceSystem(String id) throws NotFoundException {
+    int result = repository.sourceSystemExists(id);
+    if (result > 0){
+      Instant deleted = Instant.now();
+      repository.deleteSourceSystem(id, deleted);
+    }
+    else throw new NotFoundException("Requested mapping does not exist");
```
[the remark this answers is not in the document]
Comment (southeo) [reply src/main/java/eu/dissco/orchestration/backend/service/SourceSystemService.java:62]: https://naturalis.atlassian.net/browse/DD-377?atlOrigin=eyJpIjoiN2I1ZmI3ZDc3YmM5NDk1Yjg5YzNkMTBkMTc2Zjc5MTIiLCJwIjoiaiJ9

File: src/main/java/eu/dissco/orchestration/backend/service/MappingService.java (hunk 5 above)
Comment (southeo) [reply src/main/java/eu/dissco/orchestration/backend/service/MappingService.java:54]: True true. This code was written like this originally, changed to `fetchOptionals` for both functions

**Review by samleeflang — DISMISSED** [review]

**Review by samleeflang — APPROVED** [review]

**Pull request merged**